assets/js/data/schema/selectors.js

```diff
@@ -0,0 +1,162 @@
+ * namespace, then you'd have `[ 20 ]` as the ids.  This would produce something
+ * like `/wc/blocks/products/20`
+ *
+ *
+ * @throws {Error}  If there is no route for the given arguments, then this will
+ *                  throw
+ *
+ * @return {string} The route if it is available.
+ */
+export const getRoute = createRegistrySelector(
+	( select ) => ( state, namespace, modelName, ids = [] ) => {
+		const hasResolved = select( STORE_KEY ).hasFinishedResolution(
+			'getRoutes',
+			[ namespace ]
+		);
+		state = state.routes;
+		let error = '';
+		if ( ! state[ namespace ] ) {
+			error = sprintf(
+				'There is no route for the given namespace (%s) in the store',
+				namespace
+			);
+		} else if ( ! state[ namespace ][ modelName ] ) {
+			error = sprintf(
+				'There is no route for the given model name (%s) in the store',
+				modelName
+			);
+		}
+		if ( error !== '' ) {
+			if ( hasResolved ) {
+				throw new Error( error );
+			}
+			return '';
+		}
+		const route = getRouteFromModelEntries(
+			state[ namespace ][ modelName ],
+			ids
+		);
+		if ( route === '' ) {
+			if ( hasResolved ) {
+				throw new Error(
+					sprintf(
+						'While there is a route for the given namespace (%s) and model name (%s), there is no route utilizing the number of ids you included in the select arguments. The available routes are: (%s)',
+						namespace,
+						modelName,
+						JSON.stringify( state[ namespace ][ modelName ] )
+					)
+				);
+			}
+		}
+		return route;
+	}
+);
+
+/**
+ * Return all the routes for a given namespace.
+ *
+ * @param {Object} state     The current state.
+ * @param {string} namespace The namespace to return routes for.
+ *
+ * @return {Array} An array of all routes for the given namespace.
+ */
+export const getRoutes = createRegistrySelector(
+	( select ) => ( state, namespace ) => {
+		const hasResolved = select( STORE_KEY ).hasFinishedResolution(
+			'getRoutes',
+			[ namespace ]
+		);
+		const routes = state.routes[ namespace ];
+		if ( ! routes ) {
+			if ( hasResolved ) {
+				throw new Error(
+					sprintf(
+						'There is no route for the given namespace (%s) in the store',
+						namespace
+					)
+				);
+			}
+			return [];
+		}
+		let namespaceRoutes = [];
+		for ( const modelName in routes ) {
+			namespaceRoutes = [
+				...namespaceRoutes,
+				...Object.keys( routes[ modelName ] ),
+			];
+		}
+		return namespaceRoutes;
+	}
+);
+
+/**
+ * Returns the route from the given slice of the route state.
+ *
+ * @param {Object} stateSlice This will be a slice of the route state from a
+ *                            given namespace and model name.
+ * @param {Array} [ids=[]]  Any id references that are to be replaced in
+ *                            route placeholders.
+ *
+ * @returns {string}  The route or an empty string if nothing found.
+ */
+const getRouteFromModelEntries = ( stateSlice, ids = [] ) => {
+	// convert to array for easier discovery
+	stateSlice = Object.entries( stateSlice );
```

Comment:
Looks are deceiving :) 

The reason why I'm using `Object.entries` is because although we only care about the values when deciding if there's a match, we need the key from the actual matched array:

```
const [ matchingRoute, routePlaceholders ] = match || [];
```

Reply:
With that said, thanks to your comment, I did spot where I can simplify the conditional (see 1406eb2277)